controllers/control/k8ssandratask_controller.go

```diff
@@ -0,0 +1,375 @@
+			// First time we've noticed the deletion, clean up dependents and remove the finalizer
+			if err := r.deleteCassandraTasks(ctx, kTask, kc, logger); err != nil {
+				return ctrl.Result{}, err
+			}
+			logger.Info("Removing finalizer")
+			controllerutil.RemoveFinalizer(kTask, k8ssandraTaskFinalizer)
+			if err := r.Update(ctx, kTask); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Handle TTL
+	if kTask.Status.CompletionTime != nil {
+		timeNow := metav1.Now()
+		deletionTime := calculateDeletionTime(kTask)
+		if deletionTime.IsZero() { // No TTL set, we're done
+			return ctrl.Result{}, nil
+		}
+
+		if deletionTime.Before(timeNow.Time) {
+			logger.Info("Deleting K8ssandraTask due to expired TTL")
+			err := r.Delete(ctx, kTask)
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+
+		// Reschedule for later deletion
+		nextRunTime := deletionTime.Sub(timeNow.Time)
+		logger.Info("Scheduling K8ssandraTask deletion", "Delay", nextRunTime)
+		return ctrl.Result{RequeueAfter: nextRunTime}, nil
+	}
+
+	// Create CassandraTasks and/or gather their status
+	if !kcExists {
+		return r.reportInvalidSpec(ctx, kTask, "Unknown K8ssandraCluster %s.%s", kcKey.Namespace, kcKey.Name)
+	}
+	patch := client.MergeFrom(kTask.DeepCopy())
+	if dcs, err := filterDcs(kc, kTask.Spec.Datacenters); err != nil {
+		return r.reportInvalidSpec(ctx, kTask, err.Error())
+	} else {
+		for _, dc := range dcs {
+			dcNamespace := utils.FirstNonEmptyString(dc.Meta.Namespace, kc.Namespace)
+			desiredCTask := newCassandraTask(kTask, dcNamespace, dc.Meta.Name)
+
+			remoteClient, err := r.ClientCache.GetRemoteClient(dc.K8sContext)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			cTaskKey := client.ObjectKeyFromObject(desiredCTask)
+			actualCTask := &cassapi.CassandraTask{}
+			if err = remoteClient.Get(ctx, cTaskKey, actualCTask); err != nil {
+				if errors.IsNotFound(err) {
+					if err = remoteClient.Create(ctx, desiredCTask); err != nil {
+						return ctrl.Result{}, err
+					}
+					r.recordCassandraTaskCreated(kTask, desiredCTask, dc.K8sContext)
+				} else {
+					return ctrl.Result{}, err
+				}
+			} else {
+				logger.Info("CassandraTask already exists, retrieving its status", "CassandraTask", cTaskKey)
+				if kTask.Status.Datacenters == nil {
+					kTask.Status.Datacenters = make(map[string]cassapi.CassandraTaskStatus)
+				}
+				kTask.Status.Datacenters[dc.Meta.Name] = actualCTask.Status
+			}
+		}
+		wasComplete := kTask.Status.CompletionTime != nil
+		kTask.BuildGlobalStatus()
+		if err = r.Status().Patch(ctx, kTask, patch); err != nil {
+			return ctrl.Result{}, err
+		}
+		// If the status update set a completion time, we want to reconcile again in order to handle the TTL. But
+		// because we configured GenerationChangedPredicate in SetupWithManager(), we ignore our own status updates.
+		// Requeue manually just for this case.
+		if !wasComplete && kTask.Status.CompletionTime != nil {
```

Comment:
This is confusing. Let me open this if:

``if !(kTask.Status.CompletionTime != nil) && kTask.Status.CompletionTime != nil {`` .. if CompletionTime is nil and isn't nil, lets take this path. 

That can't be right?

Reply:
`kTask.BuildGlobalStatus()` can change the value.

Reply:
Do we really need to know if it has changed status here? Now the Status().Update(ctx) is called in any case, even if the status didn't change. Do we have a reason to not requeue if it is completed?

Reply:
Ok, my idea was to not requeue if the time was already set, but actually if that's the case we never get this far, the `if kTask.Status.CompletionTime != nil` near the top of the method always returns. So you're right, the boolean is not needed.